src/ui/reducers/hitCounts.ts

```diff
@@ -0,0 +1,180 @@
+// change any time either of those factors change.
+export const getCacheKeyForSourceHitCounts = (
+  state: UIState,
+  sourceId: string,
+  lineNumber: number = 0
+) => {
+  const sourceDetails = getSourceDetails(state, sourceId);
+  if (!sourceDetails) {
+    throw `Source with ID ${sourceId} not found`;
+  }
+
+  const correspondingSourceIdsSection = sourceDetails.correspondingSourceIds.join("&");
+  const { lower, upper } = getBoundsForLineNumber(lineNumber);
+  const lineNumberSection = `${lower}-${upper}`;
+  const rangeSection = getCacheKeyForFocusRegion(state);
+
+  return [correspondingSourceIdsSection, lineNumberSection, rangeSection].join("|");
+};
+
+export const fetchHitCounts = (sourceId: string, lineNumber: number): UIThunkAction => {
+  return async (dispatch, getState, { ThreadFront }) => {
+    const cacheKey = getCacheKeyForSourceHitCounts(getState(), sourceId, lineNumber);
+    const status = hitCountsSelectors.selectById(getState(), cacheKey)?.status;
+
+    if (status === LoadingState.LOADING || status === LoadingState.LOADED) {
+      return;
+    }
+
+    try {
+      dispatch(hitCountsRequested(cacheKey));
+      const { lower, upper } = getBoundsForLineNumber(lineNumber);
+      const locations = await ThreadFront.getBreakpointPositionsCompressed(sourceId);
+      const locationsToFetch = locations.filter(
+        location => location.line >= lower && location.line < upper
+      );
+      // When you fetch possible breakpoints, you will receive a list of columns
+      // for each breakable line. We only display the number of times that the
+      // *first* possible breakpoint on a line was hit, so we can throw the rest
+      // of them away.
+      locationsToFetch.map(location => {
+        const sortedColumns = sortBy(location.columns, (a, b) => a - b);
+        return {
+          ...location,
+          columns: sortedColumns.slice(0, 1),
+        };
+      });
+      const focusRegion = getFocusRegion(getState());
+      const range = focusRegion ? rangeForFocusRegion(focusRegion) : undefined;
+      const correspondingSourceIds = getCorrespondingSourceIds(getState(), sourceId)!;
+
+      const hitCounts = await fetchProtocolHitCounts(
+        correspondingSourceIds,
+        locationsToFetch,
+        range
+          ? {
+              beginPoint: range.begin.point,
+              endPoint: range.end.point,
+            }
+          : null
+      );
+      dispatch(hitCountsReceived({ id: cacheKey, hitCounts }));
+    } catch (e) {
+      dispatch(hitCountsFailed({ id: cacheKey, error: String(e) }));
+      throw e;
+    }
+  };
+};
+
+export const { hitCountsRequested, hitCountsReceived, hitCountsFailed } = hitCountsSlice.actions;
+
+export const getHitCountsForSource = (state: UIState, sourceId: string) => {
+  const cacheKey = getCacheKeyForSourceHitCounts(state, sourceId);
+  return (
+    hitCountsAdapter.getSelectors().selectById(state.hitCounts.hitCounts, cacheKey)?.hitCounts ||
```

Comment:
:hand: You can move the `getSelectors()` call out. We only need to do that once, then reuse the generated selector inside here.

Reply:
oh shoot, thanks for the catch, I thought I eliminated all of these already but this one snuck by me.